content/200-concepts/100-components/01-prisma-schema/07-referential-actions.mdx

```diff
@@ -0,0 +1,315 @@
+    authorId  Int
+}
+
+model User {
+  id    Int         @id @default(autoincrement())
+  posts Post[]
+}
+```
+
+##### Result of using `NoAction`?
+
+If a `User` has a related post in the `Post` table you will **not** be able to delete them. This will result in a runtime error.
+
+### SetNull
+
+The `SetNull` action will set relation [scalar fields](./data-model#scalar-fields) to null if the relation is updated or deleted. If one or more of the relations scalar fields are required a runtime error will be thrown.
+
+```prisma file=schema.prisma highlight=5;add|4;delete
+model Post {
+    id        Int     @id @default(autoincrement())
+    title     String
+    author    User    @relation(fields: [authorId], references: [id])
+    author    User    @relation(fields: [authorId], references: [id], onDelete: SetNull)
+    authorId  Int
+}
+
+model User {
+  id    Int         @id @default(autoincrement())
+  posts Post[]
+}
+```
+
+##### Result of using `SetNull`?
+
+If a `User` is deleted then the corresponding `authorId` field in the `Post` table will be set to `Null`
+
+### SetDefault
+
+The `SetDefault` action will set relation [scalar fields](./data-model#scalar-fields) to their default values upon updated or deletion of a related record. If no defaults are provided for any of the scalar fields, a runtime error will be thrown.
+
+```prisma file=schema.prisma highlight=5;add|4;delete
+model Post {
+    id        Int     @id @default(autoincrement())
+    title     String
+    author    User    @relation(fields: [authorId], references: [id])
+    author    User    @relation(fields: [authorId], references: [id], onDelete: SetDefault)
+    authorId  Int
+}
+
+model User {
+  id    Int         @id @default(autoincrement())
+  posts Post[]
+}
+```
+
+##### Result of using `SetDefault`?
+
+If a `User` is deleted then the corresponding `authorId` field in the `Post` table will be set to its default value.
+
+
+## Upgrade guides
+
+The following lists a number of ways to upgrade depending on your scenario.
+
+1. Upgrading Prisma to 2.26.0 _**without**_ adding the feature flag.
+
+As the preview feature is opt in, you should notice **no** difference to your database, its constraints or how it handles relations.
+
+2. Upgrading Prisma to 2.26.0 and _**enabling**_ the feature flag without adding custom referential actions. (Using the **new** defaults)
+
+If you want to keep your current database constraints but upgrade you can enable the feature flag. To do so you must have a required relation in your schema, then run either a Prisma Migration or `db push`, this will change the default for `ON DELETE` from `CASCADE` TO `RESTRICT`.
+
+The change is relevant for both the Prisma Client _and_ Migration. If an attempt is made to delete a referenced object, if their are related records, a database error will be thrown.
+
+3. Upgrading Prisma to 2.26.0, _**enabling**_ the feature flag and **adding custom referential actions**.
```

Comment:
Maybe here would make sense to mention introspection before modifying the data model, so the user would get the old default referential actions written to it before starting to do changes...

Reply:
Do you mean they run an introspection to get the **new** defaults before updating their data models?

Reply:
So, let's imagine you have this data model:

```prisma
model A {
  id Int @id
  bs B[]
}

model B {
  id Int @id
  a_id Int
  a A @relation(fields: [a_id], references: [id])
}
```

This will with the old system do something like:

```sql
CREATE TABLE A (id int PRIMARY KEY);
CREATE TABLE B (
    id int PRIMARY KEY, 
    a_id int NOT NULL,
    CONSTRAINT foo FOREIGN KEY (a_id) REFERENCES A(id) ON DELETE CASCADE ON UPDATE CASCADE
)
```

Now, we modify our data model with the generator block:

```prisma
generator js {
  provider = "prisma-client-js"
  previewFeatures = ["referentialActions"]
}
```

Now, the previous data model would have different defaults, so just as a plain migration it would be like:

```sql
CREATE TABLE A (id int PRIMARY KEY);
CREATE TABLE B (
    id int PRIMARY KEY, 
    a_id int NOT NULL,
    CONSTRAINT foo FOREIGN KEY (a_id) REFERENCES A(id) ON DELETE RESTRICT ON UPDATE CASCADE
)
```

See the different `ON DELETE`!

So, if you modify the data model and add the preview feature, but instead of migrating you run `prisma introspect`, you'll get a bit different data model:

```prisma
model A {
  id Int @id
  bs B[]
}

model B {
  id Int @id
  a_id Int
  a A @relation(fields: [a_id], references: [id], onDelete: Cascade)
}
```

So, if you migrate after setting the preview feature, you'll get a migration that alters the foreign key to have the `ON DELETE RESTRICT` action. If you instead run `prisma introspect`, the data model will be having the `onDelete: Cascade` due to the old action not being the same as the new default.